adapter/memquota2/memquota.go

```diff
@@ -0,0 +1,241 @@
+
+	if args.QuotaAmount > 0 {
+		return h.alloc(instance, args, q)
+	} else if args.QuotaAmount < 0 {
+		args.QuotaAmount = -args.QuotaAmount
+		return h.free(instance, args, q)
+	}
+	return adapter.QuotaResult2{}, nil
+}
+
+func (h *handler) alloc(instance *quota.Instance, args adapter.QuotaRequestArgs, q config.Params_Quota) (adapter.QuotaResult2, error) {
+	amount, exp, err := h.common.handleDedup(instance, args, func(key string, currentTime time.Time, currentTick int64) (int64, time.Time,
+		time.Duration) {
+		result := args.QuotaAmount
+
+		// we optimize storage for non-expiring quotas
+		if q.ValidDuration == 0 {
+			inUse := h.cells[key]
+
+			if result > q.MaxAmount-inUse {
+				if !args.BestEffort {
+					return 0, time.Time{}, 0
+				}
+
+				// grab as much as we can
+				result = q.MaxAmount - inUse
+			}
+			h.cells[key] = inUse + result
+			return result, time.Time{}, 0
+		}
+
+		window, ok := h.windows[key]
+		if !ok {
+			seconds := int32((q.ValidDuration + time.Second - 1) / time.Second)
+			window = newRollingWindow(q.MaxAmount, int64(seconds)*ticksPerSecond)
+			h.windows[key] = window
+		}
+
+		if !window.alloc(result, currentTick) {
+			if !args.BestEffort {
+				return 0, time.Time{}, 0
+			}
+
+			// grab as much as we can
+			result = window.available()
+			_ = window.alloc(result, currentTick)
+		}
+
+		return result, currentTime.Add(q.ValidDuration), q.ValidDuration
+	})
+
+	return adapter.QuotaResult2{
+		Status:        status.OK,
+		Amount:        amount,
+		ValidDuration: exp,
+	}, err
+}
+
+func (h *handler) free(instance *quota.Instance, args adapter.QuotaRequestArgs, q config.Params_Quota) (adapter.QuotaResult2, error) {
+	amount, _, err := h.common.handleDedup(instance, args, func(key string, currentTime time.Time, currentTick int64) (int64, time.Time,
+		time.Duration) {
+		result := args.QuotaAmount
+
+		if q.ValidDuration == 0 {
+			inUse := h.cells[key]
+
+			if result >= inUse {
+				// delete the cell since it contains no useful state
+				delete(h.cells, key)
+				return inUse, time.Time{}, 0
+			}
+
+			h.cells[key] = inUse - result
+			return result, time.Time{}, 0
+		}
+
+		// WARNING: Releasing quota in the case of rate limits is
+		//          inherently racy. A release can easily end up
+		//          freeing quota in the wrong window.
+
+		window, ok := h.windows[key]
+		if !ok {
+			return 0, time.Time{}, 0
+		}
+
+		result = window.release(result, currentTick)
+
+		if window.available() == q.MaxAmount {
+			// delete the cell since it contains no useful state
+			delete(h.windows, key)
+		}
+
+		return result, time.Time{}, 0
+	})
+
+	return adapter.QuotaResult2{
+		Status: status.OK,
+		Amount: amount,
+	}, err
+}
+
+func (h *handler) Close() error {
+	h.common.ticker.Stop()
+	return nil
+}
+
+////////////////// Bootstrap //////////////////////////
+
+// GetBuilderInfo returns the BuilderInfo associated with this adapter implementation.
+func GetBuilderInfo() adapter.BuilderInfo {
```

Comment:
Nit: move this to the top of the file. It's effectively the summary of what's going on in this adapter, I think the whole file reads better with this up top.

Reply:
This follows the layout of all the other adapters I created. I was going to suggest changing your adapters to move GetBuilderInfo to the end of the file :-)

I find this bit of boilerplate not that interesting which is why I put it at the end, out of the way. I was actually thinking about moving this out into a standard info.go file to further isolate it...

Reply:
I'll move my stuff down to the bottom I guess :)